### ISSUE TITLE
OutOfMemoryError while generating PDFs thumbnails using internal conversion

### ISSUE THREAD

**Issue by lfcnassif** [sepinf-inc/IPED#1032]
An user is having OOME issues and sent this trace:
2022-03-22 17:06:44	[ERROR]	[gpinf.indexer.IndexFiles]			Processing Error: 
```
java.lang.Exception: Worker-26 Error while processing /item 3.E01/vol_vol2/xxxxx/xxxxxx.zip>>xxxxxx.rar>>xxxx/xxxxxxxxxxxx.plt (6128057bytes)
	at dpf.sp.gpinf.indexer.process.Worker.process(Worker.java:181) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.Worker.run(Worker.java:233) ~[iped-engine-3.18.12.jar:?]
Caused by: java.lang.OutOfMemoryError: Java heap space
	at java.util.Arrays.copyOf(Unknown Source) ~[?:1.8.0_321]
	at java.lang.AbstractStringBuilder.ensureCapacityInternal(Unknown Source) ~[?:1.8.0_321]
	at java.lang.AbstractStringBuilder.append(Unknown Source) ~[?:1.8.0_321]
	at java.lang.StringBuilder.append(Unknown Source) ~[?:1.8.0_321]
	at dpf.sp.gpinf.indexer.util.TextCache.write(TextCache.java:58) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.task.ParsingTask.safeProcess(ParsingTask.java:388) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.task.ParsingTask.process(ParsingTask.java:330) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.task.AbstractTask.processMonitorTimeout(AbstractTask.java:221) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:159) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.task.AbstractTask.sendToNextTask(AbstractTask.java:201) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:167) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.task.AbstractTask.sendToNextTask(AbstractTask.java:201) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:167) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.task.AbstractTask.sendToNextTask(AbstractTask.java:201) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:167) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.task.AbstractTask.sendToNextTask(AbstractTask.java:201) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:167) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.task.AbstractTask.sendToNextTask(AbstractTask.java:201) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:167) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.task.AbstractTask.sendToNextTask(AbstractTask.java:201) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:167) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.task.AbstractTask.sendToNextTask(AbstractTask.java:201) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:167) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.task.AbstractTask.sendToNextTask(AbstractTask.java:201) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:167) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.task.AbstractTask.sendToNextTask(AbstractTask.java:201) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:167) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.task.AbstractTask.sendToNextTask(AbstractTask.java:201) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:167) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.task.AbstractTask.sendToNextTask(AbstractTask.java:201) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:167) ~[iped-engine-3.18.12.jar:?]
	at dpf.sp.gpinf.indexer.process.task.AbstractTask.sendToNextTask(AbstractTask.java:201) ~[iped-engine-3.18.12.jar:?]
```
He has 24GB of heap and enableExternalParsing = true. Looking at the full log, I saw he enabled the DocThumbs module and I saw other OOMEs while parsing some PDFs. But we don't log if pdfThumbs = internal/external, this could be the cause.
I asked him to change that option and to send the generated heap dump to be sure.

**Comment by lfcnassif** [comment]
I've just put a paragraph about DocThumbs option in our wiki:
https://github.com/sepinf-inc/IPED/wiki/Troubleshooting#outofmemoryerror

**Comment by lfcnassif** [comment]
Looking at the heap dump, I've just confirmed several GBs are used by PDFBox cache/objects created by DocThumbsTask. I'll make  a small change and ask the user to run processing again with internal docThumbs enabled.

**Comment by lfcnassif** [comment]
One strange thing is that ~6GB of heap was being used by PDFBox ScratchFile objects, other GBs seems to be used by the documents resources cache.

**Comment by tc-wleite** [comment]
> Looking at the heap dump, I've just confirmed several GBs are used by PDFBox cache/objects created by DocThumbsTask. I'll make a small change and ask the user to run processing again with internal docThumbs enabled.

You meant external, right?

**Comment by lfcnassif** [comment]
> You meant external, right?

"internal". I've already suggested him to use "external" to workaround the issue before analyzing the dump. Now I asked him to run again with "internal" to test the disabled resource cache I've pushed for testing, when the "external" processing ends.

If disabling resource cache helps avoid OOME, ideally we should run some tests to evaluate the performance impact...

**Comment by lfcnassif** [comment]
Actually the internal/external PDF conversion config is already logged, I looked for the wrong log message. Updating this issue title to a better description...

**Comment by lfcnassif** [comment]
> One strange thing is that ~6GB of heap was being used by PDFBox ScratchFile objects

Tooking a closer look at the dump, this is caused because there are 608 ScratchFile objects (and other 605 PDDocument & 605 DefaulktResourceCache objects), consuming up to 16MB each. This is caused because we are using `Executors.newCachedThreadPool()` and when timeouts happen (there are many of them in the log) the background thread is interrupted, but this could not end the running pdfbox code if the library doesn't check for interrupts, so we could end up with many background threads running consuming memory...

If we change to `Executors.newFixedThreadPool()` this can cause a deadlock if previous submitted pdf thumbs Runnables don't end...

I'll check if forcibly closing the opened PDDocument object helps, but this may not work in some cases...

Other option would be to change the default config to "external". The performance could be improved keeping PDF thumbs processes pools instead of starting a new one for every pdf, but this would need some effort of course and I'm sure this was left as a future improvement, I'll open a new ticket to track it.

**Comment by lfcnassif** [comment]
PS: using external PDF conversion, there is no rendering contention, leading to a few normal timeouts.

**Comment by lfcnassif** [comment]
Fixed by 73115a4

**Comment by THausherr** [comment]
It would be interesting to get the file that has the stack trace containing "PixelToParallelogramConverter". This looks familiar in the context of extremely long rendering times.

**Comment by lfcnassif** [comment]
Thanks @THausherr! I'll try to find it next week and check with the user who sent it if we could share.

**Comment by lfcnassif** [comment]
User just confirmed that running pdf to thumb conversion using external processes also fixed the issue.

**Comment by lfcnassif** [comment]
Hi @THausherr. I've just found and checked the triggering file. It seems to contain some vector graphics related to an engineering project. Chrome and PDFBox take about 5 seconds to render the file (Adobe Reader takes about half the time), that seems reasonable to me. I think the issue here is related to some kind of low level synchronization in (Oracle?) Java 8 when rendering dozens of files in parallel, that doesn't happen with java 11, as explained by @tc-wleite in details on #119.

**Comment by lfcnassif** [comment]
So, this doesn't seem related to a specific file, but when many of them are rendered at the same time in (Oracle) java 8.

**Comment by aberenguel** [comment]
I'm suffering with OOME when parsing PDF files, using the latest master commit.
I caught the hprof file and analyzed in JProfiler.  Almost all memory (24G) is consumed by int[] arrays from PDFBox cache.Each array has about 20MB.

A picture of JProfiler
![image](https://user-images.githubusercontent.com/2973359/163527343-ef622e9a-06a6-4440-bd0c-aae5c4cc4ab9.png)

**Comment by THausherr** [comment]
It would be interesting to have the file.

**Comment by lfcnassif** [comment]
Hi @aberenguel, could you continue to expand that tree to check if the triggering module is ParsingTask or DocThumbTask? Enabling `externalParsing` config in ParsingTaskConfig should workaround the issue, like described in our troubleshooting guide.

**Comment by aberenguel** [comment]
Unfortunately I will be able to check it on Monday (I've lost the access to my remote machine).
As a first try I set `pdfThumbs = external`, but I haven't seen the result yet.

**Comment by lfcnassif** [comment]
Thanks for testing.  But if the issue is parsing, that config won't help. With latest master, setting `externalParsing = true` is enough to workaround both scenarios, since pdf thumbs is also sent to external parsing processes after #1034 was merged. It is also a lot faster than `pdfThumbs = external` as detailed on #1034.

**Comment by aberenguel** [comment]
> externalParsing

I went deeper in the references of int[] arrays and got these thread stack traces:

![image](https://user-images.githubusercontent.com/2973359/163854730-d9431c44-c0d8-4be9-aea5-b807c155e6b8.png)



```
java stack of ParsingThread-434
    at org.apache.tika.parser.pdf.PDFParser.parse(java.io.InputStream, org.xml.sax.ContentHandler, org.apache.tika.metadata.Metadata, org.apache.tika.parser.ParseContext) (line: 177)
    at dpf.sp.gpinf.indexer.parsers.PDFOCRTextParser.parse(java.io.InputStream, org.xml.sax.ContentHandler, org.apache.tika.metadata.Metadata, org.apache.tika.parser.ParseContext) (line: 178)
    at org.apache.tika.parser.CompositeParser.parse(java.io.InputStream, org.xml.sax.ContentHandler, org.apache.tika.metadata.Metadata, org.apache.tika.parser.ParseContext) (line: 280)
    at dpf.sp.gpinf.indexer.parsers.IndexerDefaultParser.parse(java.io.InputStream, org.xml.sax.ContentHandler, org.apache.tika.metadata.Metadata, org.apache.tika.parser.ParseContext) (line: 246)
    at dpf.sp.gpinf.indexer.io.ParsingReader$BackgroundParsing.run() (line: 247)
    at java.util.concurrent.Executors$RunnableAdapter.call() (line: 515)
    at java.util.concurrent.FutureTask.run() (line: 264)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(java.util.concurrent.ThreadPoolExecutor$Worker) (line: 1128)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run() (line: 628)
    at java.lang.Thread.run() (line: 829)
```

```
java stack of ParsingThread-434
    at org.apache.tika.parser.pdf.PDF2XHTML.endPage(org.apache.pdfbox.pdmodel.PDPage) (line: 137)
    at org.apache.tika.parser.pdf.PDF2XHTML$AngleDetectingPDF2XHTML.processPage(org.apache.pdfbox.pdmodel.PDPage) (line: 277)
    at org.apache.tika.parser.pdf.AbstractPDF2XHTML.processPages(org.apache.pdfbox.pdmodel.PDPageTree) (line: 986)
    at org.apache.pdfbox.text.PDFTextStripper.writeText(org.apache.pdfbox.pdmodel.PDDocument, java.io.Writer) (line: 269)
    at org.apache.tika.parser.pdf.PDF2XHTML.process(org.apache.pdfbox.pdmodel.PDDocument, org.xml.sax.ContentHandler, org.apache.tika.parser.ParseContext, org.apache.tika.metadata.Metadata, org.apache.tika.parser.pdf.PDFParserConfig) (line: 96)
    at org.apache.tika.parser.pdf.PDFParser.parse(java.io.InputStream, org.xml.sax.ContentHandler, org.apache.tika.metadata.Metadata, org.apache.tika.parser.ParseContext) (line: 177)
    at dpf.sp.gpinf.indexer.parsers.PDFOCRTextParser.parse(java.io.InputStream, org.xml.sax.ContentHandler, org.apache.tika.metadata.Metadata, org.apache.tika.parser.ParseContext) (line: 178)
    at org.apache.tika.parser.CompositeParser.parse(java.io.InputStream, org.xml.sax.ContentHandler, org.apache.tika.metadata.Metadata, org.apache.tika.parser.ParseContext) (line: 280)
    at dpf.sp.gpinf.indexer.parsers.IndexerDefaultParser.parse(java.io.InputStream, org.xml.sax.ContentHandler, org.apache.tika.metadata.Metadata, org.apache.tika.parser.ParseContext) (line: 246)
    at dpf.sp.gpinf.indexer.io.ParsingReader$BackgroundParsing.run() (line: 247)
    at java.util.concurrent.Executors$RunnableAdapter.call() (line: 515)
    at java.util.concurrent.FutureTask.run() (line: 264)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(java.util.concurrent.ThreadPoolExecutor$Worker) (line: 1128)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run() (line: 628)
    at java.lang.Thread.run() (line: 829)
```

```
java stack of ParsingThread-434
    at org.apache.tika.parser.pdf.ImageGraphicsEngine.run() (line: 128)
    at org.apache.tika.parser.pdf.PDF2XHTML.extractImages(org.apache.pdfbox.pdmodel.PDPage) (line: 157)
    at org.apache.tika.parser.pdf.PDF2XHTML.endPage(org.apache.pdfbox.pdmodel.PDPage) (line: 137)
    at org.apache.tika.parser.pdf.PDF2XHTML$AngleDetectingPDF2XHTML.processPage(org.apache.pdfbox.pdmodel.PDPage) (line: 277)
    at org.apache.tika.parser.pdf.AbstractPDF2XHTML.processPages(org.apache.pdfbox.pdmodel.PDPageTree) (line: 986)
    at org.apache.pdfbox.text.PDFTextStripper.writeText(org.apache.pdfbox.pdmodel.PDDocument, java.io.Writer) (line: 269)
    at org.apache.tika.parser.pdf.PDF2XHTML.process(org.apache.pdfbox.pdmodel.PDDocument, org.xml.sax.ContentHandler, org.apache.tika.parser.ParseContext, org.apache.tika.metadata.Metadata, org.apache.tika.parser.pdf.PDFParserConfig) (line: 96)
    at org.apache.tika.parser.pdf.PDFParser.parse(java.io.InputStream, org.xml.sax.ContentHandler, org.apache.tika.metadata.Metadata, org.apache.tika.parser.ParseContext) (line: 177)
    at dpf.sp.gpinf.indexer.parsers.PDFOCRTextParser.parse(java.io.InputStream, org.xml.sax.ContentHandler, org.apache.tika.metadata.Metadata, org.apache.tika.parser.ParseContext) (line: 178)
    at org.apache.tika.parser.CompositeParser.parse(java.io.InputStream, org.xml.sax.ContentHandler, org.apache.tika.metadata.Metadata, org.apache.tika.parser.ParseContext) (line: 280)
    at dpf.sp.gpinf.indexer.parsers.IndexerDefaultParser.parse(java.io.InputStream, org.xml.sax.ContentHandler, org.apache.tika.metadata.Metadata, org.apache.tika.parser.ParseContext) (line: 246)
    at dpf.sp.gpinf.indexer.io.ParsingReader$BackgroundParsing.run() (line: 247)
    at java.util.concurrent.Executors$RunnableAdapter.call() (line: 515)
    at java.util.concurrent.FutureTask.run() (line: 264)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(java.util.concurrent.ThreadPoolExecutor$Worker) (line: 1128)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run() (line: 628)
    at java.lang.Thread.run() (line: 829)
```

**Comment by aberenguel** [comment]
It seems to be related to pdfbox cache in `PDFOCRTextParser`.

**Comment by lfcnassif** [comment]
It seems strange to me different stacktraces been shown by JProfiler coming from the same thread in the hprof file... I usually use eclipse to analyze heap dumps.

Anyway, if the issue comes from parsing module, turning `enableExternalParsing` on should workaround the issue.

And if you can find the/one triggering file like @THausherr said and if it could be shared, that would help to fix the issue in upstream project.

**Comment by aberenguel** [comment]
I've seen that the error is triggered when it starts to process a folder with a lot of PDFs. I will try to isolated the problematic PDF file.

**Comment by aberenguel** [comment]
> It would be interesting to have the file.

@THausherr, I've got the file that triggers this memory issue. How can I send it to you?

**Comment by THausherr** [comment]
@aberenguel ideally just drop it here; if you don't want that, send it to  tilman AT snafu DOT de, and mention what level of confidentiality you need . If it is too big, upload it to a sharehoster and send me the address.
